pkg/controller/alicloudmachineclass.go

```diff
@@ -67,6 +67,15 @@ func (c *controller) machineToAlicloudMachineClassDelete(obj interface{}) {
 	}
 }
 
+func (c *controller) machineToAlicloudMachineClassUpdate(oldObj, newObj interface{}) {
+	c.machineToAlicloudMachineClassAdd(oldObj)
+	c.machineToAlicloudMachineClassAdd(newObj)
```

Comment:
why would we want to reconcile both older and newer versions? 🤔

Reply:
This function only receives the machine object, and we need to enqueue the machineClass object. And if the `kind` of the machineClass has changed (possible with migration) we might need to check both objects. 

Refer - https://github.com/gardener/machine-controller-manager/blob/ce211984c40181b626ee7e5417e3e2f2aa6dab97/pkg/controller/alicloudmachineclass.go#L66.

Reply:
How about the following:

```
if new.Spec.Class.Kind ==  old.Spec.Class.Kind == Ali {
         c.machineToAlicloudMachineClassAdd(newObj)
} else {
         if new.Spec.Class.Kind == Ali {
              c.machineToAlicloudMachineClassAdd(newObj)
         } else old.Spec.Class.Kind == Ali {
              c.machineToAlicloudMachineClassAdd(oldObj)
         }
}
```
This would prevent unnecessary double operations in non-migration scenarios which is majority of the times. WDYT?

Reply:
Good suggestion @ggaurav10 . How about this code keeping your logic in mind?

```suggestion
	oldMachine, ok := oldObj.(*v1alpha1.Machine)
	if oldMachine == nil || !ok {
		klog.Warningf("Couldn't get machine from object: %+v", oldObj)
		return
	}
	newMachine, ok := oldObj.(*v1alpha1.Machine)
	if newMachine == nil || !ok {
		klog.Warningf("Couldn't get machine from object: %+v", newObj)
		return
	}

	if oldMachine.Spec.Class.Kind == newMachine.Spec.Class.Kind {
		if newMachine.Spec.Class.Kind == AWSMachineClassKind {
			// Both old and new machine refer to the same machineClass object
			// And the correct kind so enqueuing only one of them.
			c.awsMachineClassQueue.Add(newMachine.Spec.Class.Name)
		}
	} else {
		// If both are pointing to different machineClasses
		// we might have to enqueue both.
		if oldMachine.Spec.Class.Kind == AWSMachineClassKind {
			c.awsMachineClassQueue.Add(oldMachine.Spec.Class.Name)
		}
		if newMachine.Spec.Class.Kind == AWSMachineClassKind {
			c.awsMachineClassQueue.Add(newMachine.Spec.Class.Name)
		}
	}
```

Reply:
lgtm 👍

Reply:
And of course will replace the AWSMachineClassKind with class kind and queue with class queue

Reply:
Have made the suggested change here - https://github.com/gardener/machine-controller-manager/pull/531/commits/91db3fff2956dc2e0a56c52dc42ce8998b1330c4.  PTAL. 

cc: @hardikdr

Reply:
```
newMachine, ok := oldObj.(*v1alpha1.Machine)
```
In the suggestion, above, shouldn't it be newObj ?

Reply:
And you mean ClassKind above in the suggestion right, confirming though you already mentioned:
```
		if newMachine.Spec.Class.Kind == AWSMachineClassKind {
			c.awsMachineClassQueue.Add(newMachine.Spec.Class.Name)
		}
```

Reply:
>In the suggestion, above, shouldn't it be newObj ?
Yes

Reply:
Fixed it here - https://github.com/gardener/machine-controller-manager/pull/531/commits/e6ce7d31a660a9ba10973718c5b4c0d7aceb5686#diff-e37e335b3785e3c4f506cae9bde4ff92c2d1c968e78b9dba2a4bdf0cad3b228aR76.